DataFormats/Track/interface/alpaka/PixelTrackUtilities.h

```diff
@@ -0,0 +1,198 @@
+#ifndef DataFormats_Track_PixelTrackUtilities_h
+#define DataFormats_Track_PixelTrackUtilities_h
+
+#include "Geometry/CommonTopologies/interface/SimplePixelTopology.h"
+#include "DataFormats/Track/interface/PixelTrackDefinitions.h"
+#include "DataFormats/Track/interface/PixelTrackLayout.h"
+
+// Methods that operate on View and ConstView of the TrackSoA, and cannot be class methods.
+template <typename TrackerTraits>
+struct TracksUtilities {
+  using TrackSoAView = typename TrackSoA<TrackerTraits>::template TrackSoAHeterogeneousLayout<>::View;
+  using TrackSoAConstView = typename TrackSoA<TrackerTraits>::template TrackSoAHeterogeneousLayout<>::ConstView;
+  using hindex_type = typename TrackSoA<TrackerTraits>::hindex_type;
+
+  // State at the Beam spot
+  // phi,tip,1/pt,cotan(theta),zip
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static constexpr float charge(const TrackSoAConstView &tracks, int32_t i) {
```

Comment:
Unfortunately with this approach the compiler fails the template deduction.
Maybe we can find a smarter workaround...

Reply:
This seems to work, but maybe it's too generic (until we have concepts) ?
```c++
  template <typename TrackSoAViewT>
  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static constexpr float charge(TrackSoAViewT const& tracks, int32_t i) {
    float v = tracks[i].state()(2);
    return float((0.0f < v) - (v < 0.0f));
  }
```

Reply:
> This seems to work, but maybe it's too generic (until we have concepts) ?

I agree it would be too generic, and if kept in `struct TracksUtilities` (which is what I guess from the use of `static`), it would not help in simplifying the function calls.

Maybe better to be left to be improved later at this point.

Reply:
Add to https://github.com/cms-sw/cmssw/issues/43796#issuecomment-1912011615 for a follow up.

Reply:
After spending some time searching online and doing some trial and error, the only solution I found is to explicitly specialise the three traits:
```
  template <typename T>
  struct IsTrackSoAConstView : std::false_type {};
  template <>
  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::Phase1>> : std::true_type {};
  template <>
  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::Phase2>> : std::true_type {};
  template <>
  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::HIonPhase1>> : std::true_type {};
  
  template <typename T>
  constexpr bool isTrackSoAConstView = IsTrackSoAConstView<T>::value;
  
  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static constexpr float charge(ConstView const& tracks, int32_t i) {
    //was: std::copysign(1.f, tracks[i].state()(2)). Will be constexpr with C++23
    float v = tracks[i].state()(2);
    return float((0.0f < v) - (v < 0.0f));
  } 
```